### PR TITLE
[FLINK-31664][table]Add ARRAY_INTERSECT function

### PR DESCRIPTION
**Pull request by hanyuzheng7**
## What is the purpose of the change
This is an implementation of ARRAY_INTERSECT function
Returns an array of the elements in the intersection of array1 and array2, without duplicates.

## Brief change log
ARRAY_INTERSECT for Table API and SQL

## Syntax:

`ARRAY_INTERSECT(array1, array2)`

## Arguments:
array: An ARRAY to be handled.

## Returns: 
Returns an array of the elements in the intersection of array1 and array2, without duplicates.

## Examples:

```
SELECT array_intersect(array(1, 2, 3), array(1, 3, 5));
[1,3]
```

## Verifying this change
CollectionFunctionsITCase

## see also:
spark https://spark.apache.org/docs/latest/api/sql/index.html#array_intersect
presto https://prestodb.io/docs/current/functions/array.html

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / no)
  - The serializers: (yes / no / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / no / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / no / don't know)
  - The S3 file system connector: (yes / no / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by hanyuzheng7** [comment]
@flinkbot run azure

**Comment by hanyuzheng7** [comment]
@flinkbot run azure

**Comment by hanyuzheng7** [comment]
@flinkbot run azure

**Comment by hanyuzheng7** [comment]
@flinkbot run azure

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/BuiltInFunctionDefinitions.java [hunk 1]
```diff
@@ -388,7 +388,16 @@ ANY, and(logical(LogicalTypeRoot.BOOLEAN), LITERAL)
                     .inputTypeStrategy(sequence(ANY))
                     .outputTypeStrategy(nullableIfArgs(explicit(DataTypes.INT().notNull())))
                     .runtimeProvided()
-                    .internal()
```
Comment (dawidwys) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/BuiltInFunctionDefinitions.java:391]: Why did you remove the `internal` here?

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ArrayIntersectFunction.java [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+                if (element == null) {
+                    isNullPresentInArrayOne = true;
+                } else {
+                    ObjectContainer objectContainer = new ObjectContainer(element);
+                    if (seen.contains(objectContainer)) {
+                        list.add(element);
+                    }
+                }
+            }
+            if (isNullPresentInArrayTwo && isNullPresentInArrayOne) {
```
Comment (dawidwys) [inline flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ArrayIntersectFunction.java:108]: The same question as in the other PR. Shouldn't we care about the order within the first array?

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ArrayIntersectFunction.java [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+                } else {
+                    ObjectContainer objectContainer = new ObjectContainer(element);
+                    seen.add(objectContainer);
+                }
+            }
+            boolean isNullPresentInArrayOne = false;
+            for (int pos = 0; pos < array1.size(); pos++) {
+                final Object element = elementGetter.getElementOrNull(array1, pos);
+                if (element == null) {
+                    isNullPresentInArrayOne = true;
```
Comment (dawidwys) [inline flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ArrayIntersectFunction.java:100]: Can't we just check if it was present in the second array and add it to the `list`?

**Review by dawidwys — CHANGES_REQUESTED** [review]

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/BuiltInFunctionDefinitions.java (hunk 1 above)
Comment (hanyuzheng7) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/BuiltInFunctionDefinitions.java:391]: It seems that I removed it during solve conflict. I will add it back.

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ArrayIntersectFunction.java (hunk 2 above)
Comment (hanyuzheng7) [reply flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ArrayIntersectFunction.java:108]: Yes, we can care about the order. I will change it after array_except merged.

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ArrayIntersectFunction.java (hunk 3 above)
Comment (hanyuzheng7) [reply flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ArrayIntersectFunction.java:100]: Do you mean NULL? We need to check whether both array1 and array2 contain NULL. If both contain NULL, we can include it in the list.

The description of array_intersect is like this
`Returns an array of the elements in the intersection of array1 and array2, without duplicates.`

**Comment by liuyongvs** [comment]
hi @hanyuzheng7 @dawidwys ， i submit pr before, why submit another pr?
https://github.com/apache/flink/pull/22629
 https://github.com/apache/flink/pull/23959

**Comment by dawidwys** [comment]
Hey @liuyongvs I am sorry for the situation we ended up here. You're right we should've checked with you. I admit I have not been careful enough here. Please accept my apologies.

Let's try to solve it now somehow @liuyongvs @hanyuzheng7 @zzzzzzzs
1. How do you all feel if @liuyongvs takes care of `ARRAY_INTERSECT`? Both his and @hanyuzheng7 PRs are closest to be completed.
2. @hanyuzheng7 continues with `ARRAY_EXCEPT` which can reuse parts of the solution from `ARRAY_INTERSECT`

Do all fell okish with such a plan? I know it's not a perfect solution, but I want to find a way out of it.

**Comment by hanyuzheng7** [comment]
@dawidwys @liuyongvs Ok

**Comment by liuyongvs** [comment]
hi @dawidwys will you help review this https://github.com/apache/flink/pull/24526
